File: servant/src/CardsProject/Tournaments/TournamentService.hs
{-# LANGUAGE OverloadedStrings #-}
module CardsProject.Tournaments.TournamentService
  ( validateTournament, start, cancel, complete, generate_round, calculate_prize_distribution, register_player, is_full, enumToText, assertTransition, allowedTransitions, transitionDraftToRegistration, transitionRegistrationToOngoing, transitionRegistrationToCancelled, transitionOngoingToCompleted, transitionOngoingToCancelled, transitionCompletedToDraft, transitionCancelledToDraft
  ) where

import CardsProject.Tournaments.Types
import Control.Exception (throwIO)
import System.IO.Error (userError)
import Data.Text (Text)
import Data.Maybe (fromMaybe)
import qualified Data.Text
import Database.SQLite.Simple
import Database.SQLite.Simple.FromField ()
import CardsProject.Db (withDb)

-- Domain service for Tournament
validateTournament :: NewTournament -> Either String NewTournament
validateTournament body
  | bTournamentMaxPlayers body < 2 = Left "max_players: must be >= 2"
  | bTournamentMaxPlayers body > 512 = Left "max_players: must be <= 512"
  | not ((bTournamentMaxPlayers body >= 2 && bTournamentMaxPlayers body <= 512)) = Left "Tournament must allow between 2 and 512 players"
  | not (bTournamentEntryFee body >= 0) = Left "Entry fee must not be negative"
  | not (bTournamentPrizePool body >= 0) = Left "Prize pool must not be negative"
  | otherwise = validateTournamentImplies body

validateTournamentImplies :: NewTournament -> Either String NewTournament
validateTournamentImplies body
  | (bTournamentEndTime body /= Nothing) && not (maybe True (> bTournamentStartTime body) (bTournamentEndTime body)) = Left "End time must be after start time"
  | otherwise = Right body

-- @invoke behavior stub (no-op)
start :: Int -> IO ()
start _eid = throwIO (userError "start not implemented")

-- @invoke behavior stub (no-op)
cancel :: Int -> IO ()
cancel _eid = throwIO (userError "cancel not implemented")

-- @invoke behavior stub (no-op)
complete :: Int -> IO ()
complete _eid = throwIO (userError "complete not implemented")

-- @invoke behavior stub (no-op)
generate_round :: Int -> IO ()
generate_round _eid = throwIO (userError "generate_round not implemented")

-- @invoke behavior stub (no-op)
calculate_prize_distribution :: Int -> IO Text
calculate_prize_distribution _eid = throwIO (userError "calculate_prize_distribution not implemented")

-- @invoke behavior stub (no-op)
register_player :: Int -> IO ()
register_player _eid = throwIO (userError "register_player not implemented")

-- @invoke behavior stub (no-op)
is_full :: Int -> IO Bool
is_full _eid = throwIO (userError "is_full not implemented")

-- ── Lifecycle state machine ─────────────────────────────────────────
allowedTransitions :: [(Text, [Text])]
allowedTransitions =
  [   ("Draft", ["Registration"])
  ,  ("Registration", ["Ongoing", "Cancelled"])
  ,  ("Ongoing", ["Completed", "Cancelled"])
  ]

-- Convert status enum to Text: FooStatusType_Active -> "Active"
enumToText :: Show a => a -> Text
enumToText v = Data.Text.pack $ drop 1 $ dropWhile (/= '_') (show v)

assertTransition :: Text -> Text -> IO ()
assertTransition current to_ = do
  let allowed = maybe [] id (lookup current allowedTransitions)
  if to_ `elem` allowed
    then return ()
    else throwIO (userError $ "Transition " ++ show current ++ " -> " ++ show to_ ++ " not allowed")

transitionDraftToRegistration :: Int -> IO Tournament
transitionDraftToRegistration eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      assertTransition (enumToText (tournamentStatus record)) "Registration"
      execute conn "UPDATE tournaments SET status = ? WHERE id = ?" ("Registration" :: Text, eid)
      updated <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
      case updated of
        (r:_) -> return r
        []    -> throwIO (userError "Tournament not found after update")

transitionRegistrationToOngoing :: Int -> IO Tournament
transitionRegistrationToOngoing eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      assertTransition (enumToText (tournamentStatus record)) "Ongoing"
      execute conn "UPDATE tournaments SET status = ? WHERE id = ?" ("Ongoing" :: Text, eid)
      start eid  -- @after
      updated <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
      case updated of
        (r:_) -> return r
        []    -> throwIO (userError "Tournament not found after update")

transitionRegistrationToCancelled :: Int -> IO Tournament
transitionRegistrationToCancelled eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      assertTransition (enumToText (tournamentStatus record)) "Cancelled"
      execute conn "UPDATE tournaments SET status = ? WHERE id = ?" ("Cancelled" :: Text, eid)
      cancel eid  -- @after
      updated <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
      case updated of
        (r:_) -> return r
        []    -> throwIO (userError "Tournament not found after update")

transitionOngoingToCompleted :: Int -> IO Tournament
transitionOngoingToCompleted eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      assertTransition (enumToText (tournamentStatus record)) "Completed"
      execute conn "UPDATE tournaments SET status = ? WHERE id = ?" ("Completed" :: Text, eid)
      complete eid  -- @after
      calculate_prize_distribution eid  -- @after
      updated <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
      case updated of
        (r:_) -> return r
        []    -> throwIO (userError "Tournament not found after update")

transitionOngoingToCancelled :: Int -> IO Tournament
transitionOngoingToCancelled eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      assertTransition (enumToText (tournamentStatus record)) "Cancelled"
      execute conn "UPDATE tournaments SET status = ? WHERE id = ?" ("Cancelled" :: Text, eid)
      cancel eid  -- @after
      updated <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
      case updated of
        (r:_) -> return r
        []    -> throwIO (userError "Tournament not found after update")

transitionCompletedToDraft :: Int -> IO Tournament
transitionCompletedToDraft eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      throwIO (userError "Transition Completed -> Draft is not allowed")

transitionCancelledToDraft :: Int -> IO Tournament
transitionCancelledToDraft eid = withDb $ \conn -> do
  rows <- query conn "SELECT id, public_id, name, description, status, bracket_data, format, tournament_type, max_players, entry_fee, prize_pool, start_time, end_time, is_online, location, rules_text, created_at, season_id, organizer_id FROM tournaments WHERE id = ?" (Only eid) :: IO [Tournament]
  case rows of
    [] -> throwIO (userError "Tournament not found")
    (record:_) -> do
      throwIO (userError "Transition Cancelled -> Draft is not allowed")

-- ── Lifecycle hooks ─────────────────────────────────────────────────

-- TODO: implement sync_season_stats
syncSeasonStatsHook :: a -> IO ()
syncSeasonStatsHook _ = return ()

-- TODO: implement prevent_delete_if_ongoing
preventDeleteIfOngoingHook :: a -> IO ()
preventDeleteIfOngoingHook _ = return ()

File: luminus/src/cards_project/tournaments/tournament_service.clj
(ns cards_project.tournaments.tournament-service
  (:require [next.jdbc :as jdbc]
            [next.jdbc.result-set :as rs]
            [cards_project.tournaments.tournament-queries :as queries]
            [cards_project.db :refer [db-spec]]))

(defn validate-tournament
  "Validate and transform params before persistence."
  [params]
  params)

; ── Domain behavior stubs ──────────────────────────────────────────
(defn- start-behavior! [id]
  ; TODO: implement start
  nil)

(defn- cancel-behavior! [id]
  ; TODO: implement cancel
  nil)

(defn- complete-behavior! [id]
  ; TODO: implement complete
  nil)

(defn- generate-round-behavior! [id]
  ; TODO: implement generate_round
  nil)

(defn- calculate-prize-distribution-behavior! [id]
  ; TODO: implement calculate_prize_distribution
  nil)

(defn- register-player-behavior! [id player-id deck-id]
  ; TODO: implement register_player
  nil)

(defn- is-full-behavior! [id]
  ; TODO: implement is_full
  nil)

; ── Lifecycle state machine ─────────────────────────────────────────
(def ^:private allowed-transitions
  {   "Draft" ["Registration"]
   "Registration" ["Ongoing", "Cancelled"]
   "Ongoing" ["Completed", "Cancelled"]})

(defn- assert-transition! [current to]
  (let [allowed (get allowed-transitions current [])]
    (when-not (some #(= % to) allowed)
      (throw (ex-info (str "Transition " current " -> " to " not allowed")
                      {:status 409})))))

(defn transition-draft-to-registration!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (do
      (assert-transition! (get record :status) "Registration")
      (when (nil? (get record :name))
        (throw (ex-info "name is required for Draft -> Registration" {:status 422})))
      (when (nil? (get record :start_time))
        (throw (ex-info "start_time is required for Draft -> Registration" {:status 422})))
      (jdbc/execute-one! db-spec
        ["UPDATE tournaments SET status = ? WHERE id = ?" "Registration" id])
      (queries/get-tournament-by-id db-spec {:id id}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-registration-to-ongoing!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (do
      (assert-transition! (get record :status) "Ongoing")
      (jdbc/execute-one! db-spec
        ["UPDATE tournaments SET status = ? WHERE id = ?" "Ongoing" id])
      (start-behavior! id) ; @after
      (queries/get-tournament-by-id db-spec {:id id}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-registration-to-cancelled!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (do
      (assert-transition! (get record :status) "Cancelled")
      (jdbc/execute-one! db-spec
        ["UPDATE tournaments SET status = ? WHERE id = ?" "Cancelled" id])
      (cancel-behavior! id) ; @after
      (queries/get-tournament-by-id db-spec {:id id}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-ongoing-to-completed!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (do
      (assert-transition! (get record :status) "Completed")
      (jdbc/execute-one! db-spec
        ["UPDATE tournaments SET status = ? WHERE id = ?" "Completed" id])
      (complete-behavior! id) ; @after
      (calculate-prize-distribution-behavior! id) ; @after
      (queries/get-tournament-by-id db-spec {:id id}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-ongoing-to-cancelled!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (do
      (assert-transition! (get record :status) "Cancelled")
      (jdbc/execute-one! db-spec
        ["UPDATE tournaments SET status = ? WHERE id = ?" "Cancelled" id])
      (cancel-behavior! id) ; @after
      (queries/get-tournament-by-id db-spec {:id id}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-completed-to-draft!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (throw (ex-info "Transition Completed -> Draft is not allowed" {:status 409}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn transition-cancelled-to-draft!
  [id]
  (if-let [record (queries/get-tournament-by-id db-spec {:id id})]
    (throw (ex-info "Transition Cancelled -> Draft is not allowed" {:status 409}))
    (throw (ex-info "Tournament not found" {:id id :status 404}))))

(defn start!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (start-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn cancel!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (cancel-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn complete!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (complete-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn generate-round!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (generate-round-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn calculate-prize-distribution!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (calculate-prize-distribution-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn register-player!
  [id player-id deck-id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (register-player-behavior! id player-id deck-id)
    (throw (ex-info "Tournament not found" {:id id}))))

(defn is-full!
  [id]
  (if (queries/get-tournament-by-id db-spec {:id id})
    (is-full-behavior! id)
    (throw (ex-info "Tournament not found" {:id id}))))

